File: webclient/javascript/models/person.js
import Model from './model';
import Position from '../position';

class Person extends Model {
  constructor(id, name, position, inventory) {
    super();
    this.id = id;
    this.name = name;
    this.position = position;
    this.inventory = inventory;
  }

  static get UPDATE_POSITION() { return 'update_position'; }
  static get UPDATE_INVENTORY() { return 'update_inventory'; }
  static get REMOVE() { return 'remove'; }
  static get STEP_SIZE() { return 0.00001; }

  static unpack(person) {
    return new Person(
      person.id,
      person.name,
      person.position,
      person.inventory
    );
  }

  updatePosition(position) {
    this.position = position;
    this.notify(Person.UPDATE_POSITION, position);
  }

  remove() {
    this.notify(Person.REMOVE);
  }

  addToInventory(items) {
    for (const name in items) {
      if (this.inventory[name]) {
        this.inventory[name] += items[name];
      } else {
        this.inventory[name] = 1;
      }
    }
    this.notify(Person.UPDATE_INVENTORY);
  }

  removeFromInventory(items) {
    for (const name in items) {
      if (this.inventory[name] && this.inventory[name] - items[name] >= 0) {
        this.inventory[name] -= items[name];
      } 
    }
    this.notify(Person.UPDATE_INVENTORY);
  }

  static stepTowards(initialPosition, finalPosition) {
    const latDiff = finalPosition.lat - initialPosition.lat;
    const lngDiff = finalPosition.lng - initialPosition.lng;
    const distance = Position.distance(initialPosition, finalPosition);
    return {
      lat: Person.STEP_SIZE * latDiff / distance,
      lng: Person.STEP_SIZE * lngDiff / distance
    };
  }

  static isClose(position1, position2) {
    return Position.distance(position1, position2) < Person.STEP_SIZE;
  }
}

export default Person;

File: server/models/person.py
from collections import defaultdict
from math import sqrt

class Person:
    STEP_SIZE = 0.00001

    def __init__(self, person_id, name, position):
        self.id = person_id
        self.name = name
        self.position = position
        self.active = True
        self.inventory = defaultdict(int)

    def pack(self):
        return {
            'id': self.id,
            'name': self.name,
            'position': self.position,
            'inventory': self.inventory
        }

    def deactivate(self):
        self.active = False

    def move_direction(self, direction):
        self.position = _move_direction[direction](self.position)

    def move_step(self, step):
        self.position = make_step(
            position=self.position,
            lat=step.get('lat'),
            lng=step.get('lng')
        )

    def canTakeResource(self, resource):
        return close(self.position, resource.get('position'))

    def addToInventory(self, items):
        for name, value in items.iteritems():
            self.inventory[name] += value




def make_step(position, lat=0, lng=0):
    return {
        'lat': position.get('lat') + lat,
        'lng': position.get('lng') + lng,
    }

def distance(position1, position2):
    latDiff = position1.get('lat') - position2.get('lat')
    lngDiff = position1.get('lng') - position2.get('lng')
    return sqrt(latDiff*latDiff + lngDiff*lngDiff)

def close(position1, position2):
    return distance(position1, position2) < 0.00009


_move_direction = {
    'up': lambda position: make_step(position, lat=Person.STEP_SIZE),
    'down': lambda position: make_step(position, lat=-Person.STEP_SIZE),
    'right': lambda position: make_step(position, lng=Person.STEP_SIZE),
    'left': lambda position: make_step(position, lng=-Person.STEP_SIZE)
}
